server/matchmaker/algorithm/team_matchmaker.py

```diff
@@ -0,0 +1,292 @@
+        """
+        If we simply fetch the highest/lowest rated single player search we may overshoot our
+        goal to get the most balanced teams, so we try them all until we find the one that brings
+        us closest to the rating average i.e. balanced teams
+        If there is no single player search we have hit a search combination that is impossible to
+        separate into two teams e.g. (3, 3, 2) for 4v4
+        """
+        team_avg = search.average_rating
+        if not searches_dict[1]:
+            self._logger.warning("given searches are impossible to split in even teams because of party sizes")
+            raise UnevenTeamsException
+        candidates = SortedList(searches_dict[1], key=lambda s: s.cumulated_rating)
+
+        if avg - team_avg < 0:
+            iterator = iter(candidates)
+        else:
+            iterator = iter(candidates[::-1])
+        candidate = next(iterator)
+        for item in iterator:
+            team_avg = get_average_rating([search, candidate])
+            avg_delta = abs(avg - team_avg)
+            new_team_avg = get_average_rating([search, item])
+            new_avg_delta = abs(avg - new_team_avg)
+            self._logger.debug("delta with %s is %s (avg is %s)", [candidate], avg_delta, team_avg)
+            self._logger.debug("new delta with %s is %s (avg is %s)", [item], new_avg_delta, new_team_avg)
+            if new_avg_delta > avg_delta:
+                break
+            candidate = item
+        self._logger.debug("used %s as filler", [candidate])
+        return candidate
+
+    def calculate_game_quality(self, match: Match) -> Game:
+        newbie_bonus = 0
+        time_bonus = 0
+        ratings = []
+        for team in match:
+            time_bonus += team.failed_matching_attempts * config.TIME_BONUS
+            time_bonus = min(time_bonus, config.MAXIMUM_TIME_BONUS)
+            newbie_bonus += team.has_newbie() * config.NEWBIE_BONUS
+            for search in team.get_original_searches():
+                ratings.append(search.average_rating)
+
+        rating_disparity = abs(match[0].cumulated_rating - match[1].cumulated_rating)
+        fairness = max((config.MAXIMUM_RATING_IMBALANCE - rating_disparity) / config.MAXIMUM_RATING_IMBALANCE, 0)
+        deviation = statistics.pstdev(ratings)
+        uniformity = max((config.MAXIMUM_RATING_DEVIATION - deviation) / config.MAXIMUM_RATING_DEVIATION, 0)
+
+        quality = fairness * uniformity + newbie_bonus + time_bonus
+        self._logger.debug(
+            "bonuses: %s rating disparity: %s -> fairness: %f deviation: %f -> uniformity: %f -> game quality: %f",
+            newbie_bonus + time_bonus, rating_disparity, fairness, deviation, uniformity, quality)
+        return Game(match, quality)
+
+    def _pick_best_noncolliding_games(self, games: List[Game]) -> List[Match]:
+        for game in list(games):
+            if game.quality < config.MINIMUM_GAME_QUALITY:
+                games.remove(game)
```

Comment:
Since you're copying `games` anyways, you might as well do this in a way that doesn't modify the original. Plus I believe this way is faster.
```suggestion
        games = [game for game in games if game.quality >= config.MINIMUM_GAME_QUALITY]
```